packages/Discretization/src/DTK_Interpolation_def.hpp

```diff
@@ -0,0 +1,669 @@
+                imported_cell_indices, imported_points, imported_query_ids,
+                topo, topo_id, _filtered_points[topo_id],
+                _filtered_cell_indices[topo_id], _filtered_query_ids[topo_id],
+                _point_indices_map[topo_id], _reference_points[topo_id],
+                _point_in_cell[topo_id] );
+}
+
+template <typename DeviceType>
+void Interpolation<DeviceType>::computeNodeOffset(
+    unsigned int const n_cells,
+    Kokkos::View<DTK_CellTopology *, DeviceType> cell_topo_view,
+    Kokkos::View<unsigned int[DTK_N_TOPO], DeviceType> n_nodes_per_topo,
+    Kokkos::View<unsigned int *, DeviceType> nodes_per_cell,
+    Kokkos::View<unsigned int *, DeviceType> node_offset )
+{
+    using ExecutionSpace = typename DeviceType::execution_space;
+    Kokkos::parallel_for( REGION_NAME( "fill_nodes_per_cell" ),
+                          Kokkos::RangePolicy<ExecutionSpace>( 0, n_cells ),
+                          KOKKOS_LAMBDA( int const i ) {
+                              nodes_per_cell( i ) =
+                                  n_nodes_per_topo( cell_topo_view( i ) );
+                          } );
+    Kokkos::fence();
+    Kokkos::parallel_scan(
+        REGION_NAME( "compute_node_offset" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n_cells ),
+        KOKKOS_LAMBDA( int i, int &update, bool final_pass ) {
+            if ( final_pass )
+                node_offset( i ) = update;
+            update += nodes_per_cell( i );
+        } );
+    Kokkos::fence();
+}
+
+template <typename DeviceType>
+void Interpolation<DeviceType>::fillCellDofIds(
+    unsigned int const topo_id, unsigned int const n_cells,
+    unsigned int const n_nodes_per_topo,
+    Kokkos::View<DTK_CellTopology *, DeviceType> cell_topo_view,
+    Kokkos::View<unsigned int *, DeviceType> offset,
+    Kokkos::View<unsigned int *, DeviceType> node_offset,
+    Kokkos::View<LocalOrdinal *, DeviceType> object_dof_ids,
+    Kokkos::View<LocalOrdinal **, DeviceType> cell_dofs_ids )
+{
+    using ExecutionSpace = typename DeviceType::execution_space;
+    Kokkos::parallel_for(
+        REGION_NAME( "fill_cell_dof_ids" + std::to_string( topo_id ) ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n_cells ),
+        KOKKOS_LAMBDA( int const i ) {
+            if ( cell_topo_view( i ) == topo_id )
+            {
+                unsigned int const k = offset( i );
+                for ( unsigned int node = 0; node < n_nodes_per_topo; ++node )
+                {
+                    unsigned int const n = node_offset( i ) + node;
+                    cell_dofs_ids( k, node ) = object_dof_ids( n );
+                }
+            }
+        } );
+    Kokkos::fence();
+}
+
+template <typename DeviceType>
+void Interpolation<DeviceType>::computeTopologies()
+{
+    // This needs to be done on the host because we will use _n_cells_per_topo
+    // to allocate Kokkos::View and this needs to be done on the host
```

Comment:
Yeah, we could. The reason why `_n_cells_per_topo` is not a Kokkos View is because it is too small to work efficiently on a GPU. However, since we need to copy `_cell_topologies_view` on the cpu, we probably don't gain anything.